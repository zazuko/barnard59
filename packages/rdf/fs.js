import { Transform } from 'readable-stream'
import fromFile from 'rdf-utils-fs/fromFile.js'

export function parse() {
  return new Transform({
    objectMode: true,
    transform: async function (path, encoding, callback) {
      try {
        const fileStream = fromFile(path)

        let failed = false
        fileStream.on('data', quad => this.push(quad))
        fileStream.on('error', (e) => {
          callback(e)
          failed = true
        })
        fileStream.on('end', () => !failed && callback())
      } catch (e) {
        callback(e)
      }
    },
  })
}
