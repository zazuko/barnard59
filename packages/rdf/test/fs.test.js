import { Readable } from 'stream'
import { resolve } from 'path'
import * as url from 'url'
import $rdf from 'barnard59-env'
import chai, { expect } from 'chai'
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot'
import fromStream from 'rdf-dataset-ext/fromStream.js'
import toCanonical from 'rdf-dataset-ext/toCanonical.js'
import { parse as unbound } from '../fs.js'

const __dirname = url.fileURLToPath(new URL('.', import.meta.url))

const parse = unbound.bind({ env: $rdf })

describe('fs', () => {
  chai.use(jestSnapshotPlugin())

  describe('parse', () => {
    it('parses files from input chunks paths', async () => {
      // given
      const files = Readable.from([
        resolve(__dirname, 'support/fs-samples/A.ttl'),
        resolve(__dirname, 'support/fs-samples/B.jsonld'),
        resolve(__dirname, 'support/fs-samples/C.nq'),
      ])

      // when
      const dataset = await fromStream($rdf.dataset(), files.pipe(parse()))

      // then
      expect(toCanonical(dataset)).toMatchSnapshot()
    })

    it('fails when file does not exist', (done) => {
      // given
      const files = Readable.from([
        resolve('no-such-file.ttl'),
      ])

      // when
      const stream = files.pipe(parse())

      // then
      stream.on('error', (err) => {
        expect(err.message).to.match(/no such file or directory/)
        done()
      })
    })

    it('fails when file fails to parse', (done) => {
      // given
      const files = Readable.from([
        resolve(__dirname, 'support/fs-samples/bad.ttl'),
      ])

      // when
      const stream = files.pipe(parse())

      // then
      stream.on('error', (err) => {
        expect(err.message).to.match(/Expected entity/)
        done()
      })
    })

    it('fails when extensions is unrecognised', (done) => {
      // given
      const files = Readable.from([
        resolve(__dirname, 'support/file.unknown.extension'),
      ])

      // when
      const stream = files.pipe(parse())

      // then
      stream.on('error', (err) => {
        expect(err.message).to.match(/Unknown file extension/)
        done()
      })
    })
  })
})
