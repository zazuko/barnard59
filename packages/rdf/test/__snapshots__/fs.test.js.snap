// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fs parse parses files from input chunks paths 1`] = `
"<http://example.org/A> <http://www.w3.org/2000/01/rdf-schema#label> \\"A\\" .
<http://example.org/A> <http://www.w3.org/2000/01/rdf-schema#label> \\"a\\" .
<http://example.org/B> <http://www.w3.org/2000/01/rdf-schema#label> \\"B\\" .
<http://example.org/B> <http://www.w3.org/2000/01/rdf-schema#label> \\"b\\" .
<http://example.org/C> <http://www.w3.org/2000/01/rdf-schema#label> \\"C\\" <http://example.org/> .
<http://example.org/C> <http://www.w3.org/2000/01/rdf-schema#label> \\"c\\" <http://example.org/> .
"
`;
